### ISSUE TITLE
Upgrade to 22.0.0.11 results in 401 errors in ios app, with Android app showing stack backtrace

### ISSUE THREAD

**Issue by kpneal** [nextcloud/server#27931]
I upgraded my NextCloud instance over the weekend. The result is unusable for the iOS and Android apps. I'm seeing loads of 401 errors when I use the iOS app:

{"reqId":"RGYT8aeQ4AFQO87Uy77z","level":0,"time":"2021-07-12T00:20:14+00:00","remoteAddr":"2600:1700:e70:3b10:132:7ad:446b:a925","user":"--","app":"webdav","method":"PUT","url":"/remote.php/webdav/Photos/2021-07-09%2019:04:23%208844.jpg","message":"No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured","userAgent":"Mozilla/5.0 (iOS) Nextcloud-iOS/4.0.4","version":"22.0.0.11","exception":{"Exception":"Sabre\\DAV\\Exception\\NotAuthenticated","Message":"No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured","Code":0,"Trace":[{"file":"/usr/local/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"beforeMethod","class":"Sabre\\DAV\\Auth\\Plugin","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/usr/local/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":456,"function":"emit","class":"Sabre\\DAV\\Server","type":"->","args":["beforeMethod:PUT",[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/usr/local/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":253,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/usr/local/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":321,"function":"start","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/usr/local/www/nextcloud/apps-pkg/dav/appinfo/v1/webdav.php","line":83,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/usr/local/www/nextcloud/remote.php","line":166,"args":["/usr/local/www/nextcloud/apps-pkg/dav/appinfo/v1/webdav.php"],"function":"require_once"}],"File":"/usr/local/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Auth/Plugin.php","Line":154,"CustomMessage":"--"}}

The desktop sync client on my Mac (OS X 11.4) runs just fine. The web site runs just fine. I'm told the Android app crashes, but I don't have it in hand with details.

I'm now unable to upload any files with the iOS app. It tries, but it keeps showing the message "Waiting for upload: 401 Unauthorized". I'm pretty sure that folders not being cached is why I can see the contents of some folders but not others in the iOS app. Anything cached by the app seems just fine. Uncached, not visible.

The server was fine when it was running NC 21. I'm using php-fpm, the sqlite3 backend, and the admin's Settings/Overview page isn't showing any errors. It does warn me about using sqlite, but an install as small as mine should be fine. Files are stored in the local filesystem. I'm not using LDAP.

I did run the "php ./occ upgrade" command like normal after an upgrade of the installed software.

I'm running FreeBSD 12.2, amd64, and I'm installing software straight from the FreeBSD ports system.

Ports system installed software versions:
nextcloud-calendar-php74-2.3.0 Calendar app for Nextcloud
nextcloud-contacts-php74-4.0.0 Contacts app for Nextcloud
nextcloud-php74-22.0.0         Personal cloud which runs on your own server
nextcloud-talk-php74-12.0.0    Online meeting app for Nextcloud

apache24-2.4.48                Version 2.4.x of Apache web server
apr-1.7.0.1.6.1_1              Apache Portability Library

sqlite3-3.35.5_3,1             SQL database engine in a C library

redis-6.0.14                   Persistent key-value database with built-in net interface

php74-7.4.21                   PHP Scripting Language
php74-bcmath-7.4.21            The bcmath shared extension for php
php74-bz2-7.4.21               The bz2 shared extension for php
php74-ctype-7.4.21             The ctype shared extension for php
php74-curl-7.4.21              The curl shared extension for php
php74-dom-7.4.21               The dom shared extension for php
php74-exif-7.4.21              The exif shared extension for php
php74-fileinfo-7.4.21          The fileinfo shared extension for php
php74-filter-7.4.21            The filter shared extension for php
php74-gd-7.4.21                The gd shared extension for php
php74-gmp-7.4.21               The gmp shared extension for php
php74-iconv-7.4.21             The iconv shared extension for php
php74-intl-7.4.21              The intl shared extension for php
php74-json-7.4.21              The json shared extension for php
php74-ldap-7.4.21              The ldap shared extension for php
php74-mbstring-7.4.21          The mbstring shared extension for php
php74-opcache-7.4.21           The opcache shared extension for php
php74-openssl-7.4.21           The openssl shared extension for php
php74-pcntl-7.4.21             The pcntl shared extension for php
php74-pdo-7.4.21               The pdo shared extension for php
php74-pdo_mysql-7.4.21         The pdo_mysql shared extension for php
php74-pdo_sqlite-7.4.21        The pdo_sqlite shared extension for php
php74-pecl-APCu-5.1.20         APC User Caching
php74-pecl-imagick-3.4.4_2     PHP wrapper to the ImageMagick/GraphicsMagick library version 6
php74-pecl-redis-5.3.2         Extension to access Redis
php74-posix-7.4.21             The posix shared extension for php
php74-session-7.4.21           The session shared extension for php
php74-simplexml-7.4.21         The simplexml shared extension for php
php74-sqlite3-7.4.21           The sqlite3 shared extension for php
php74-xml-7.4.21               The xml shared extension for php
php74-xmlreader-7.4.21         The xmlreader shared extension for php
php74-xmlwriter-7.4.21         The xmlwriter shared extension for php
php74-xsl-7.4.21               The xsl shared extension for php
php74-zip-7.4.21               The zip shared extension for php
php74-zlib-7.4.21              The zlib shared extension for php

$ php ./occ app:list
Enabled:
  - accessibility: 1.7.0
  - activity: 2.15.0
  - admin_audit: 1.11.0
  - calendar: 2.3.0
  - circles: 22.0.0
  - cloud_federation_api: 1.4.0
  - comments: 1.11.0
  - contacts: 4.0.0
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - dav: 1.18.0
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_pdfviewer: 2.3.0
  - files_rightclick: 1.1.0
  - files_sharing: 1.13.2
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.11.0
  - firstrunwizard: 2.11.0
  - logreader: 2.7.0
  - lookup_server_connector: 1.9.0
  - nextcloud_announcements: 1.11.0
  - notifications: 2.10.1
  - oauth2: 1.9.0
  - password_policy: 1.12.0
  - photos: 1.4.0
  - privacy: 1.6.0
  - provisioning_api: 1.11.0
  - recommendations: 1.1.0
  - serverinfo: 1.12.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - spreed: 12.0.0
  - support: 1.5.0
  - survey_client: 1.10.0
  - systemtags: 1.11.0
  - text: 3.3.0
  - theming: 1.12.0
  - twofactor_backupcodes: 1.10.1
  - updatenotification: 1.11.0
  - user_status: 1.1.1
  - viewer: 1.6.0
  - weather_status: 1.1.0
  - workflowengine: 2.3.0
Disabled:
  - encryption
  - files_external
  - user_ldap
$

$ php ./occ config:list system
{
    "system": {
        "apps_paths": [
            {
                "path": "\/usr\/local\/www\/nextcloud\/apps",
                "url": "\/apps",
                "writable": true
            },
            {
                "path": "\/usr\/local\/www\/nextcloud\/apps-pkg",
                "url": "\/apps-pkg",
                "writable": false
            }
        ],
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "loglevel": 0,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "dbindex": 0,
            "timeout": 0
        },
        "trusted_domains": [
            "localhost",
            "share.neutralgood.org",
            "206.196.19.102",
            "[2607:f758:2280:5:9d7b:3dcd:d17c:59d1]"
        ],
        "default_phone_region": "US",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "sqlite3",
        "version": "22.0.0.11",
        "overwrite.cli.url": "http:\/\/localhost",
        "installed": true,
        "mail_smtpmode": "sendmail",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "updater.release.channel": "stable",
        "theme": ""
    }
}
$ 
Results from https://share.neutralgood.org/index.php/settings/integrity/failed
No errors have been found.

**Comment by valintw** [comment]
I suspect this issue is affecting me too. I use a couple of 3rd party apps with Nextcloud both of which have stopped working since I upgraded to 22.
They are both WebDAV clients. I use WebDAV explorer and cloudbeats media player.   WebDAV explorer works very intermittently. (Mostly not) and cloudbeats can never connect. Both give me timeout errors when trying to connect and I can see 401 errors on the Nextcloud console when they try to connect

**Comment by szaimen** [comment]
After reading the nextcloud log, I think this is a duplicate of https://github.com/nextcloud/server/issues/8011, no?

**Comment by kpneal** [comment]
The iOS issue was caused by a packaging error on the FreeBSD side. They have that fixed now. So this can't be the same as #8011. (The FreeBSD side didn't have the ".htaccess" file in the application root directory, for those wondering.)

I'm still seeing crashes from the Android client. Same stack backtrace as before. I'm not sure what I should be looking for in my logs over here. Suggestions welcome. I have the Apache log, the php-fpm logs (the one that sorta looks like a web server log, and the log where each line is in "{}" aka "curly braces"), plus a system call trace that I ran while the app was starting up and crashing.

I wish I had known up front that the iOS issue and the Android issue were different bugs. Sorry for mixing them both in the same ticket.

**Comment by Latz** [comment]
I tracked down the error to the following problem (at least in my App):

I'm always using the same function for all API calls. In this function I'm sending the following HTTP headers:

``` 
const headers = {
    'OCS-APIREQUEST': 'true',
    'User-Agent': 'myApp',
    Authorization: 'someAuthString'
  };
```
Login Flow in NC v21 didn't have a problem with receiving an "Authorization" header with meaningless content. v22 seems to evaluate the header in the Login Flow process. If I'm omitting this header everything works fine.

IMHO NC v22 is behaving correctly since the content of Authorization is correct since the client tries to authenticate using an invalid value.

**Comment by kpneal** [comment]
The issue seems fixed in 22.1.0. Closing.
